### DISCUSSION TITLE
【更新日志】2024/05/29 PaoPaoDNS 新增`EXPIRED_FLUSH`乐观缓存主动回收选项

### DISCUSSION THREAD

**Discussion by kkkgo** [kkkgo/PaoPaoDNS#154] [更新日志]
## 新增`EXPIRED_FLUSH`乐观缓存主动回收选项
[unbound](https://github.com/NLnetLabs/unbound)最新的一个[修复](https://github.com/NLnetLabs/unbound/issues/1071)解决了cachedb外部控制的问题，因此这个一直想做的功能也可以实现了。

该选项为`yes`，且在`CNAUTO`、`CNFALL`为`yes`时生效。该选项默认值为`yes`。当开启该选项时，将会主动监测递归结果中出现的乐观缓存，在乐观缓存返回后数秒后检查是否成功递归刷新了新的解析结果，如果递归失败（由两次ttl记录差值对比），将会主动回收清除该缓存。开启该选项可以有效避免乐观缓存因网络连接性不稳定而一直滞留过期记录的问题，提高DNS解析结果的实时性。

其他更新日志
- 修改递归策略，在纯递归模式下使用更积极连接更多的递归策略，在CNFALL下使用占用更低效率更高的递归策略。
- 修复某些情况下yes_only6选项不生效的问题
- 增加构建检查、debug调试信息。

**Comment by EyeShield29** [comment]
<img width="1499" alt="Screenshot 2024-05-27 at 23 29 37" src="https://github.com/kkkgo/PaoPaoDNS/assets/35490258/3a644462-3890-41d8-b60c-ef60778b0e33">
大佬好，拉取了最新镜像启动后，在图里的报错后，dns服务就挂掉了，浏览器再访问会有DNS_PROBE_FINISHED_BAD_CONFIG的报错。看着是DDL到期后重新获取失败导致的，目前先设置成把EXPIRED_FLUSH设置成no恢复了
下面是docker-compose的配置，EXPIRED_FLUSH先改为no了

**Reply by kkkgo** [reply]
修正了一个错误，等一小时后推送吧。

**Reply by kkkgo** [reply]
好了，拉一下最新镜像试试。

**Reply by EyeShield29** [reply]
拉了最新镜像，用了一段时间后，还是有图里最后那段报错信息的问题。其他配置都没调整

**Reply by kkkgo** [reply]
报错信息发一下，是不是拉的新镜像没用上

**Reply by EyeShield29** [reply]
是最新的，之前图上的info日志都没有了，有问题的时候再看日志就只有这段了，报错信息跟之前的一样
![3de8eb59-861c-41d6-90b8-8af15e26cfa1](https://github.com/kkkgo/PaoPaoDNS/assets/35490258/bf8d4ca6-77c1-42bb-82fc-52989f4216e0)

**Reply by kkkgo** [reply]
容器启动日志可以发一下吗

**Reply by EyeShield29** [reply]
[log.txt](https://github.com/kkkgo/PaoPaoDNS/files/15478880/log.txt)
报错的没有了，已经改完配置重启了，这份是EXPIRED_FLUSH设置成no的

**Reply by kkkgo** [reply]
好的，已经修复，等Actions编译一小时推送新镜像再拉取试试。

**Reply by kkkgo** [reply]
拉取最新镜像试试。

**Reply by 742038744** [reply]
大佬好，某些网站进不去，会是什么问题，比如todesk官网能进去，但是下载不了，下载页面进不去

**Reply by kkkgo** [reply]
> 大佬好，某些网站进不去，会是什么问题，比如todesk官网能进去，但是下载不了，下载页面进不去

参考方法排查：https://github.com/kkkgo/PaoPaoDNS/discussions/132#discussioncomment-9286884

**Comment by TwoBall** [comment]
flushd_server的日志是否可以关闭或者设置显示级别

**Reply by kkkgo** [reply]
默认就是没日志的（都是高于error的级别），除非你修改了mosdns.yaml

**Reply by TwoBall** [reply]
那就奇怪了，我会显示如楼上一样的info日志。![image](https://github.com/kkkgo/PaoPaoDNS/assets/9460044/50e39c7c-d0ea-44d6-b66d-c4d1be3d9af0)

**Reply by kkkgo** [reply]
确实，日志写的是开发模式，等一小时后推送编译拉新镜像试试。

**Reply by kkkgo** [reply]
拉取新镜像试试。

**Reply by TwoBall** [reply]
最新镜像没有问题

**Comment by gukey** [comment]
越来越好用了;

**Comment by Mxzjxs** [comment]
你好，这个paopaodns要如何更新呢？

**Reply by kkkgo** [reply]
更新docker镜像。

**Comment by kiss2u** [comment]
这里的乐观缓存，与adguardhome的乐观缓存是一个意思吗？

**Reply by kkkgo** [reply]
是一个意思，但只涉及递归部分。

**Reply by kiss2u** [reply]
那。。大佬，这是不是意味着可以打开adh的乐观缓存？

**Reply by kkkgo** [reply]
下游DNS的乐观缓存无法被上游控制，具体行为取决于adh，也就是你之前adh开乐观缓存是怎么样就是怎么样。

**Reply by kiss2u** [reply]
了解。。。感谢大佬解疑

**Comment by 742038744** [comment]
网络是通的，还是以前的操作方法，无论如何都拉去不下来镜像了
docker: error pulling image configuration: download failed after attempts=6: dial tcp [2a03:2880:f12c:83:face:b00c:0:25de]:443: i/o timeout.

**Reply by kkkgo** [reply]
如果你的网络环境访问Docker Hub镜像有困难，可以尝试使用public.ecr.aws镜像:

- 示例： `docker pull public.ecr.aws/sliamb/paopaodns`
- 示例： `docker run -d public.ecr.aws/sliamb/paopaodns`
